File: xssfinder-core/src/main/java/org/xssfinder/runner/WebDriverXPathFinder.java
package org.xssfinder.runner;

import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.List;

/**
 * Creates uniquely identifying XPath expressions for WebDriver WebElements
 */
class WebDriverXPathFinder {
    public String getXPath(WebElement element) {
        if (element.getAttribute("id") != null) {
            return "//" + element.getTagName().toLowerCase() + "[@id=\"" + element.getAttribute("id") + "\"]";
        } else if (element.getTagName().equalsIgnoreCase("body")) {
            return "body";
        }

        int nodeCount = 1;
        WebElement parentElement = element.findElement(By.xpath(".."));
        List<WebElement> childElements = parentElement.findElements(By.xpath("./*"));
        for (WebElement childElement : childElements) {
            if (childElement.equals(element)) {
                return getXPath(parentElement) + "/" + element.getTagName().toLowerCase() + "[" + nodeCount + "]";
            }
            if (childElement.getTagName().equalsIgnoreCase(element.getTagName())) {
                nodeCount++;
            }
        }
        throw new RuntimeException("Element is not a child of its parent");
    }
}

File: xssfinder-php-executor/src/XssFinder/Runner/WebDriverXPathFinder.php
<?php

namespace XssFinder\Runner;

use WebDriverElement;

class WebDriverXPathFinder
{
    /**
     * @param WebDriverElement $element
     * @return string
     */
    public function getXPath(WebDriverElement $element)
    {
        $elementTagName = strtolower($element->getTagName());
        if ($element->getAttribute('id')) {
            return sprintf('//%s[@id="%s"]', $elementTagName, $element->getAttribute('id'));
        } elseif ($elementTagName === 'body') {
            return 'body';
        }

        $parentElement = $element->findElement(\WebDriverBy::xpath('..'));
        $childElements = $parentElement->findElements(\WebDriverBy::xpath('./*'));
        $matchingChildNumber = 1;
        foreach ($childElements as $childElement) {
            /** @var $childElement WebDriverElement */
            if ($childElement == $element) {
                return $this->getXPath($parentElement) . sprintf('/%s[%d]', $elementTagName, $matchingChildNumber);
            } elseif (strtolower($childElement->getTagName()) === $elementTagName) {
                $matchingChildNumber++;
            }
        }
        throw new \Exception("Web driver element is not a child of its parent");
    }
}